### ISSUE TITLE
Get notified when actually started to consume messages

### ISSUE THREAD

**Issue by Timoha** [bsm/sarama-cluster#110]
Hello,

As I undestand, before I actually start receiving messages from [Messages()](https://godoc.org/github.com/bsm/sarama-cluster#Consumer.Messages) channel, there's a brief period of initialization of consumer during which I might miss some messages. Is there a way to find out exactly when consumer will start receiving messages?

I've taken a look at [Notifications()](https://godoc.org/github.com/bsm/sarama-cluster#Consumer.Notifications) channel at it looks like there might be a way to check `Claimed` field of the notifications I receive. Would that guarantee not missing any messages or there are better way to go about this? Thank you!

**Comment by dim** [comment]
I don't quite understand the question. Not missing any messages (even with multiple, distributed consumers) is the primary goal of this project. If you request a topic and set the `Consumer.Offsets.Initial` setting to `OffsetOldest` (see [here](https://github.com/Shopify/sarama/blob/e6df07fb9cf315a30d9a8d665185090d5bf84b22/config.go#L214)), it will start consuming the topic from the oldest known partition remembering offsets you [mark](https://godoc.org/github.com/bsm/sarama-cluster#Consumer.MarkOffset) as processed. The next time you start consuming the same topic using the same consumer group name, it will resume from offset where you have previously stopped. 

Hope that explains it.

**Comment by Timoha** [comment]
@dim thank you for your quick reply, let me try to clarify with my particular situation:

I have a service, which gets the latest state from the database and the updates to the state are received via kafka.
On initialization, I first subscribe to a kafka topic with `OffsetNewest` in a goroutine to not miss any updates and then start fetching the latest state from the database. Since consumer is taking its time to initialize before it actually starts getting messages from kafka (there's quite a lot of things happening before actually consuming: https://github.com/bsm/sarama-cluster/blob/master/consumer.go#L258), there's a possible race condition, which can result into inconsistent state in my service. So, I need some mechanism to get notified when consumer actually started consuming from kafka, so that I can go and fetch my latest state from storage.

I also don't mark my offsets as read because when I subscribe I only interested in latest messages since I have my state stored in the database.

**Comment by dim** [comment]
Consuming without marking offsets doesn't sound right. The goal of this
project is to allow multiple consumers to consume topics concurrently
avoiding messages accidentally being read twice.

Although I doubt that sarama cluster is the right tool for your job, you
can achieve what you're looking for through sync primitives or simple go
channels, for example:

```go
cc := make (chan struct{})
go func() {
  <-cc // wait for channel
  // now so something in the background
}()

cb := false
for m :=  range c.Messages() {
  if !cb {
    close(cc)
    cb = true
  }
}
```

**Comment by Timoha** [comment]
> Consuming without marking offsets doesn't sound right. The goal of this project is to allow multiple consumers to consume topics concurrently avoiding messages accidentally being read twice. 

Not sure to understand this, Kafka doesn't require me to commit offsets. If I have only one consumer in consumer group and if I always prefer to read from the newest offset in case I restart or fail, Kafka is flexible enough to allow that.

> Although I doubt that sarama cluster is the right tool for your job, you can achieve what you're looking for through sync primitives or simple go channels, for example:

I guess my clarification wasn't good enough because that's not exactly what I'm asking for, I can't really wait until the first message because it might arrive in a second, an hour, or a day. The issue is that between exiting from [NewConsumer](https://godoc.org/github.com/bsm/sarama-cluster#NewConsumer) and it actually starting to consume it does things like [rebalancing](https://github.com/bsm/sarama-cluster/blob/master/consumer.go#L247), [starting a heartbeat](https://github.com/bsm/sarama-cluster/blob/master/consumer.go#L254), [fetching offsets](https://github.com/bsm/sarama-cluster/blob/master/consumer.go#L502), [creating partition consumers](https://github.com/bsm/sarama-cluster/blob/master/consumer.go#L518), and only after those are done, I can be sure that I'm receiving messages (and that's where I would like to be notified).

> Although I doubt that sarama cluster is the right tool for your job

I agree with you now that `sarama-cluster` might be not right fit for me or maybe I architected my system in a wrong way, but I'll keep playing with it.

Really appreciate your help @dim!

**Comment by nsd20463** [comment]
I agree with dim. If you have one consumer in the group, and want to always restart at OffsetNewest, then a consumer group client is not what you should use. You should use the direct sarama API. Something like

  con := sarama.NewConsumer()
  part := con.ConsumePartition(topic, 0, OffsetNewest)
  for msg := range part.Messages() { ... }

except with error handling would be what you need, and it would start consuming within one RTT to the kafka brokers.
